example_scenes/advanced_tex_fonts.py

```diff
@@ -0,0 +1,46 @@
+import os
+from manim import *
+
+
+class TexTemplateFrenchCursive(TexTemplateFromFile):
+    def rebuild_cache(self):
```

Comment:
>  It is somewhat unfortunate that setting a custom template has to be done in rebuild_cache 

There is absolutely no obligation to do it in `rebuild_cache`, I would even say this is not the intended use of that method.

> but this is how TexTemplateFromFile is designed right now.

No, this is not how `TexTemplateFromFile` is designed. The class is designed to read a template file and cache its contents. 

Whoever wants to create a custom template can do so by extending the `TexTemplate` class and set up all packages they want, e.g. by changing its `CONFIG` dictionary or by just using its built-in method `append_package` like in

```python
append_package(["frcursive", ["default"]])
append_package(["mathastext", ["eulergreek","noplusnominus","noequal","nohbar","nolessnomore","noasterisk"]])
```

Reply:
Thanks for the clarification, @PhilippImhof -- and sorry for my wrong assumtion. (I admit: I basically read what `rebuild_cache` does and not much more. :upside_down_face:)

In this case the showcase example should definitely use the intended way to create a custom template; could you please modify it accordingly, @cobordism?

Reply:
No worries at all

Reply:
> In this case the showcase example should definitely use the intended way to create a custom template; could you please modify it accordingly, @cobordism?

No. I'd prefer to further refactor the TexTemplate implementation that Philipp started.

I don't think we should have an append_package function that takes latex packages and their arguments and options as arguments. Firstly it requires anyone who wants to add something to a LaTeX template to translate the line of latex into python so that this function can translate it back into latex. Why? Let's just have an append_to_preamble(txt) that inserts txt into the template verbatim.
Secondly this function is too restrictive as not all entries in the latex preamble are of the form \usepackage[a]{b}
For example, the template for ECF Skeetch includes:
```
           \usepackage[T1]{fontenc}
            \usepackage[T1]{fontenc}
            \DeclareFontFamily{T1}{fsk}{}
            \DeclareFontShape{T1}{fsk}{m}{n}{<->s*[1.315] fskmw8t}{}
            \renewcommand\rmdefault{fsk}
            \usepackage[noendash,defaultmathsizes,nohbar,defaultimath]{mathastext}
```
Other templates further require entries to the body (after \begin{document}) and not just the preamble.
For example "URW Zapf Chancery (CM Greek)", which includes:
```
            \usepackage[T1]{fontenc}
            \DeclareFontFamily{T1}{pzc}{}
            \DeclareFontShape{T1}{pzc}{mb}{it}{<->s*[1.2] pzcmi8t}{}
            \DeclareFontShape{T1}{pzc}{m}{it}{<->ssub * pzc/mb/it}{}
            \DeclareFontShape{T1}{pzc}{mb}{sl}{<->ssub * pzc/mb/it}{}
            \DeclareFontShape{T1}{pzc}{m}{sl}{<->ssub * pzc/mb/sl}{}
            \DeclareFontShape{T1}{pzc}{m}{n}{<->ssub * pzc/mb/it}{}
            \usepackage{chancery}
            \usepackage{mathastext}
            \linespread{1.05}
            \begin{document}\boldmath
```


I have started experimenting with this idea here: https://github.com/cobordism/manim/blob/refactor-tex-templates
but it is not nearly as well formatted and documented as Philipp's version and only represents an experiment of mine as to how far I could (in theory) simplify it... In this experiment I use no CONFIG whatsoever, the Tex and MathTex and related mobjects loose ther "type: text" and "type: tex" attributes as well as the "ctex" flag -- these were confusing hacks inherited from 3b1b and not good designs anyway.
As in this current PR, the templates include the desired compiler and output (eg latex/dvi or xelatex/xdv, or pdflatex/pdf). Further, I have added to the text mobjects a desired latex environment (eg none for text, align* for math mode).

In that version the example file would look like this: 
https://github.com/cobordism/manim/blob/9e7cd1ce6c37698e1105e23eda9beb224d0fbba1/example_scenes/advanced_tex_fonts.py

Reply:
I see your point.

I would really like to avoid showcasing workarounds like overriding `rebuild_cache`. If you do not agree with or simply don't like the clean solution right now, it might be best to wait with adding examples until the interface works in a more flexible way.

I also agree that it should be possible to set the entire template body, if someone knows what they are doing; the updated file from your like looks good regarding that. :-)

Reply:
> No. I'd prefer to further refactor the TexTemplate implementation that Philipp started.

Sounds like a good idea to me, especially in the context of Manim CE being much more open and dynamic than the original Manim. When I wrote the customization, I tried to remain as close as possible to the original code in order to have better chances for the new code to be merged.

> I don't think we should have an append_package function that takes latex packages and their arguments and options as arguments. Firstly it requires anyone who wants to add something to a LaTeX template to translate the line of latex into python so that this function can translate it back into latex. Why?

People who choose to use the pythonised template probably like it. Those who do not like it, will probably use the custom template file. Also, I honestly think that the overwhelming majority of user will not even use custom templates. (They might however do so, if the interface is easy enough.)

> Let's just have an append_to_preamble(txt) that inserts txt into the template verbatim.

IMHO it's not a case of either-or, it's more a situation of having one and the other. Actually, the method `append_to_preamble(self,text)` already exists.

> Secondly this function is too restrictive as not all entries in the latex preamble are of the form \usepackage[a]{b}
> For example, the template for ECF Skeetch includes:
> 
> ```
>            \usepackage[T1]{fontenc}
>             \usepackage[T1]{fontenc}
>             \DeclareFontFamily{T1}{fsk}{}
>             \DeclareFontShape{T1}{fsk}{m}{n}{<->s*[1.315] fskmw8t}{}
>             \renewcommand\rmdefault{fsk}
>             \usepackage[noendash,defaultmathsizes,nohbar,defaultimath]{mathastext}
> ```

This is a bad example, because all of this can already be done with `append_to_preamble`.

> Other templates further require entries to the body (after \begin{document}) and not just the preamble.
> For example "URW Zapf Chancery (CM Greek)", which includes:
> 
> ```
>             \usepackage[T1]{fontenc}
>             \DeclareFontFamily{T1}{pzc}{}
>             \DeclareFontShape{T1}{pzc}{mb}{it}{<->s*[1.2] pzcmi8t}{}
>             \DeclareFontShape{T1}{pzc}{m}{it}{<->ssub * pzc/mb/it}{}
>             \DeclareFontShape{T1}{pzc}{mb}{sl}{<->ssub * pzc/mb/it}{}
>             \DeclareFontShape{T1}{pzc}{m}{sl}{<->ssub * pzc/mb/sl}{}
>             \DeclareFontShape{T1}{pzc}{m}{n}{<->ssub * pzc/mb/it}{}
>             \usepackage{chancery}
>             \usepackage{mathastext}
>             \linespread{1.05}
>             \begin{document}\boldmath
> ```

That's a much better example. The solution could be `add_to_preamble` combined with the `\AtBeginDocument` macro. But having a method like `add_to_document` would be fine, of course.

> In this experiment I use no CONFIG whatsoever, the Tex and MathTex and related mobjects loose ther "type: text" and "type: tex" attributes as well as the "ctex" flag -- these were confusing hacks inherited from 3b1b and not good designs anyway.

Very well diagnosed... ;-) However, for the `ctex` flag, I do not know whether some people rely on it. OTOH, sometimes it can be necessary (and even sensible) to break backwards compatibility for the few in order to have better code for the many; I find this especially true for early releases where the "future" is much longer than the "past".

> As in this current PR, the templates include the desired compiler and output (eg latex/dvi or xelatex/xdv, or pdflatex/pdf). Further, I have added to the text mobjects a desired latex environment (eg none for text, align* for math mode).

Sounds good to me. More flexibility is always a good thing. I could even imagine that users (have the option to) specify a custom template for any `mobject` (and related), this could come in handy also for chanim (@kilacoda).